apps/confidential/confidential_test.go

```diff
@@ -805,6 +814,213 @@ func TestNewCredFromTokenProvider(t *testing.T) {
+	)
+	mockClient.AppendResponse(mock.WithBody(mock.GetTenantDiscoveryBody(lmo, "secondTenant")))
+	mockClient.AppendResponse(
+		mock.WithBody([]byte(fmt.Sprintf(`{"access_token":%q,"expires_in":%d,"refresh_in":%d,"token_type":"Bearer"}`, firstToken, expiresIn, refreshIn))),
+	)
+	// Create the client instance
+	client, err := New(fmt.Sprintf(authorityFmt, lmo, "firstTenant"), fakeClientID, cred, WithHTTPClient(&mockClient), WithInstanceDiscovery(false))
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Acquire the first token for first tenant
+	ar, err := client.AcquireTokenByCredential(context.Background(), tokenScope, WithTenantID("firstTenant"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Assert the first token is returned
+	if ar.AccessToken != firstToken {
+		t.Fatalf("wanted %q, got %q", firstToken, ar.AccessToken)
+	}
+	// Acquire the first token for second tenant
+	arSecond, err := client.AcquireTokenByCredential(context.Background(), tokenScope, WithTenantID("secondTenant"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if arSecond.AccessToken != firstToken {
+		t.Fatalf("wanted %q, got %q", firstToken, arSecond.AccessToken)
+	}
+	fixedTime := time.Now().Add(time.Duration(43400) * time.Second)
+	base.GetCurrentTime = func() time.Time {
+		return fixedTime
+	}
+	var wg sync.WaitGroup
+	done := make(chan struct{})
+	ch := make(chan error, 1)
+
+	firstTenantChecker := false
+	secondTenantChecker := false
+	mockClient.AppendResponse(
+		mock.WithBody([]byte(fmt.Sprintf(`{"access_token":%q,"expires_in":%d,"refresh_in":%d,"token_type":"Bearer"}`, secondToken+"firstTenant", expiresIn, refreshIn+44200))),
+	)
+	mockClient.AppendResponse(
+		mock.WithBody([]byte(fmt.Sprintf(`{"access_token":%q,"expires_in":%d,"refresh_in":%d,"token_type":"Bearer"}`, secondToken+"secondTenant", expiresIn, refreshIn+44200))))
+
+	for i := 0; i < 10000; i++ {
+		wg.Add(2)
+		go func() {
+			defer wg.Done()
+			ar, err := client.AcquireTokenSilent(context.Background(), tokenScope, WithTenantID("firstTenant"))
+			if err != nil {
+				select {
+				case ch <- err:
+				default:
+				}
+				return
+			}
+			if ar.AccessToken == secondToken+"firstTenant" && ar.Metadata.TokenSource == base.IdentityProvider {
+				if firstTenantChecker {
+					t.Error("Error can only call this once")
+				} else {
+					firstTenantChecker = true
+				}
+			}
+		}()
+		go func() {
+			defer wg.Done()
+			ar, err := client.AcquireTokenSilent(context.Background(), tokenScope, WithTenantID("secondTenant"))
+			if err != nil {
+				select {
+				case ch <- err:
+				default:
+				}
+				return
+			}
+			if ar.AccessToken == secondToken+"secondTenant" && ar.Metadata.TokenSource == base.IdentityProvider {
+				if secondTenantChecker {
+					t.Error("Error can only call this once")
+				} else {
+					secondTenantChecker = true
+				}
+			}
+		}()
+	}
+	// Wait for all goroutines in a separate goroutine
+	go func() {
+		wg.Wait()
+		close(done)
+		close(ch)
+	}()
```

Comment:
Why do this in a separate goroutine, and with a `done` channel? You want the goroutine executing the test to wait for the others--it can just call `wg.Wait()` as before. Maybe this will help explain the mechanics: https://play.golang.com/p/hWAE3Cenyd8

Reply:
I created the go routine, so that we can safely close the channels and create a `wait()`
I understand what i missed in there
I updated it